File: Sources/ChatUnpackCore/Domain/ChatMessage.swift
import Foundation

public struct RecognizedField: Sendable, Equatable {
  public var text: String
  public var confidence: Float
  public var isUserCorrected: Bool

  public init(text: String, confidence: Float, isUserCorrected: Bool = false) {
    self.text = text
    self.confidence = confidence
    self.isUserCorrected = isUserCorrected
  }

  public var isLowConfidence: Bool {
    confidence < 0.70 && !isUserCorrected
  }
}

public struct RecognizedLine: Identifiable, Sendable, Equatable {
  public let id: UUID
  public var text: String
  public var confidence: Float
  public var isUserCorrected: Bool

  public init(
    id: UUID = UUID(),
    text: String,
    confidence: Float,
    isUserCorrected: Bool = false
  ) {
    self.id = id
    self.text = text
    self.confidence = confidence
    self.isUserCorrected = isUserCorrected
  }

  public var isLowConfidence: Bool {
    confidence < 0.70 && !isUserCorrected
  }
}

public enum MessageKind: String, Sendable, Codable, CaseIterable {
  case text
  case image
  case voice
  case video
  case file
  case miniProgram
  case link
  case nestedRecord
  case emoji
  case unknownNonText
}

public struct ChatMessage: Identifiable, Sendable, Equatable {
  public let id: UUID
  public var sender: RecognizedField
  public var timestamp: RecognizedField
  public var body: [RecognizedLine]
  public var kind: MessageKind
  public var warnings: [ScanWarning]
  public var sourceViewportIndices: Set<Int>
  public var isPartial: Bool

  public init(
    id: UUID = UUID(),
    sender: RecognizedField,
    timestamp: RecognizedField,
    body: [RecognizedLine],
    kind: MessageKind = .text,
    warnings: [ScanWarning] = [],
    sourceViewportIndices: Set<Int> = [],
    isPartial: Bool = false
  ) {
    self.id = id
    self.sender = sender
    self.timestamp = timestamp
    self.body = body
    self.kind = kind
    self.warnings = warnings
    self.sourceViewportIndices = sourceViewportIndices
    self.isPartial = isPartial
  }

  public var hasLowConfidence: Bool {
    sender.isLowConfidence
      || timestamp.isLowConfidence
      || body.contains(where: { $0.isLowConfidence })
  }
}

File: windows/src/ChatUnpack.Core/Domain/ChatMessage.cs
namespace ChatUnpack.Core.Domain;

public record struct RecognizedField
{
  public RecognizedField(string text, double? confidence, bool isUserCorrected = false)
  {
    Text = text ?? string.Empty;
    Confidence = confidence;
    IsUserCorrected = isUserCorrected;
  }

  public string Text { get; set; }
  public double? Confidence { get; set; }
  public bool IsUserCorrected { get; set; }

  public bool IsLowConfidence => Confidence is < 0.45 && !IsUserCorrected;
}

public sealed class RecognizedLine : IEquatable<RecognizedLine>
{
  public RecognizedLine(
    string text,
    double? confidence,
    bool isUserCorrected = false,
    Guid? id = null)
  {
    Id = id ?? Guid.NewGuid();
    Text = text ?? string.Empty;
    Confidence = confidence;
    IsUserCorrected = isUserCorrected;
  }

  public Guid Id { get; }
  public string Text { get; set; }
  public double? Confidence { get; set; }
  public bool IsUserCorrected { get; set; }

  public bool IsLowConfidence => Confidence is < 0.45 && !IsUserCorrected;

  public bool Equals(RecognizedLine? other)
  {
    return other is not null
      && Id == other.Id
      && Text == other.Text
      && Confidence == other.Confidence
      && IsUserCorrected == other.IsUserCorrected;
  }

  public override bool Equals(object? obj)
  {
    return Equals(obj as RecognizedLine);
  }

  public override int GetHashCode()
  {
    return HashCode.Combine(Id, Text, Confidence, IsUserCorrected);
  }
}

public enum MessageKind
{
  Text,
  Image,
  Voice,
  Video,
  File,
  MiniProgram,
  Link,
  NestedRecord,
  Emoji,
  UnknownNonText
}

public sealed class ChatMessage : IEquatable<ChatMessage>
{
  public ChatMessage(
    RecognizedField sender,
    RecognizedField timestamp,
    IEnumerable<RecognizedLine> body,
    MessageKind kind = MessageKind.Text,
    IEnumerable<ScanWarning>? warnings = null,
    IEnumerable<int>? sourceViewportIndices = null,
    bool isPartial = false,
    Guid? id = null)
  {
    Id = id ?? Guid.NewGuid();
    Sender = sender;
    Timestamp = timestamp;
    Body = body?.ToList() ?? new List<RecognizedLine>();
    Kind = kind;
    Warnings = warnings?.ToList() ?? new List<ScanWarning>();
    SourceViewportIndices = sourceViewportIndices is null
      ? new HashSet<int>()
      : new HashSet<int>(sourceViewportIndices);
    IsPartial = isPartial;
  }

  public Guid Id { get; }
  public RecognizedField Sender { get; set; }
  public RecognizedField Timestamp { get; set; }
  public List<RecognizedLine> Body { get; set; }
  public MessageKind Kind { get; set; }
  public List<ScanWarning> Warnings { get; set; }
  public HashSet<int> SourceViewportIndices { get; set; }
  public bool IsPartial { get; set; }

  public bool HasLowConfidence => string.IsNullOrEmpty(Timestamp.Text);

  public bool Equals(ChatMessage? other)
  {
    return other is not null
      && Id == other.Id
      && Sender == other.Sender
      && Timestamp == other.Timestamp
      && Body.SequenceEqual(other.Body)
      && Kind == other.Kind
      && Warnings.SequenceEqual(other.Warnings)
      && SourceViewportIndices.SetEquals(other.SourceViewportIndices)
      && IsPartial == other.IsPartial;
  }

  public override bool Equals(object? obj)
  {
    return Equals(obj as ChatMessage);
  }

  public override int GetHashCode()
  {
    var hash = new HashCode();
    hash.Add(Id);
    hash.Add(Sender);
    hash.Add(Timestamp);
    foreach (var line in Body)
    {
      hash.Add(line);
    }

    hash.Add(Kind);
    foreach (var warning in Warnings)
    {
      hash.Add(warning);
    }

    foreach (var viewportIndex in SourceViewportIndices.OrderBy(value => value))
    {
      hash.Add(viewportIndex);
    }

    hash.Add(IsPartial);
    return hash.ToHashCode();
  }
}
